File: trabalho_3/cliente_javascript/js/services/ApiService.js
/**
 * Serviço para comunicação com a API do servidor
 */
class ApiService {
    constructor(baseUrl = 'http://localhost:8000') {
        this.baseUrl = baseUrl;
    }

    /**
     * Realiza uma requisição HTTP
     * @param {string} endpoint - Endpoint da API
     * @param {Object} options - Opções da requisição (method, body, etc.)
     * @returns {Promise<Object>} Resposta da API
     */
    async request(endpoint, options = {}) {
        const url = `${this.baseUrl}${endpoint}`;
        const config = {
            headers: {
                'Content-Type': 'application/json',
                ...options.headers
            },
            ...options
        };

        try {
            const response = await fetch(url, config);
            const data = await response.json();
            
            if (!response.ok) {
                throw new Error(data.detail || `HTTP error! status: ${response.status}`);
            }
            
            return data;
        } catch (error) {
            console.error('Erro na requisição:', error);
            throw error;
        }
    }

    // ==================== CLUBES ====================

    /**
     * Cria um novo clube
     * @param {Clube} clube - Dados do clube
     * @returns {Promise<Object>} Resposta da API
     */
    async criarClube(clube) {
        return this.request('/clubes', {
            method: 'POST',
            body: JSON.stringify(clube.toJSON())
        });
    }

    /**
     * Lista todos os clubes
     * @returns {Promise<Clube[]>} Lista de clubes
     */
    async listarClubes() {
        const data = await this.request('/clubes');
        return data.map(clubeData => Clube.fromJSON(clubeData));
    }

    /**
     * Obtém estatísticas de um clube
     * @param {string} sigla - Sigla do clube
     * @returns {Promise<Clube>} Dados do clube
     */
    async obterEstatisticasClube(sigla) {
        const data = await this.request(`/clubes/estatisticas?sigla=${sigla}`);
        return Clube.fromJSON(data);
    }

    /**
     * Atualiza um clube
     * @param {string} siglaOriginal - Sigla original do clube
     * @param {Clube} clube - Novos dados do clube
     * @returns {Promise<Object>} Resposta da API
     */
    async atualizarClube(siglaOriginal, clube) {
        return this.request(`/clubes?sigla_original=${siglaOriginal}`, {
            method: 'PUT',
            body: JSON.stringify(clube.toJSON())
        });
    }

    /**
     * Deleta um clube
     * @param {string} sigla - Sigla do clube
     * @returns {Promise<Object>} Resposta da API
     */
    async deletarClube(sigla) {
        return this.request(`/clubes?sigla=${sigla}`, {
            method: 'DELETE'
        });
    }

    /**
     * Obtém histórico de partidas de um clube
     * @param {string} sigla - Sigla do clube
     * @returns {Promise<Partida[]>} Lista de partidas
     */
    async obterHistoricoClube(sigla) {
        const data = await this.request(`/clubes/partidas?sigla=${sigla}`);
        return data.map(partidaData => Partida.fromJSON(partidaData));
    }

    // ==================== CAMPEONATOS ====================

    /**
     * Cria um novo campeonato
     * @param {Campeonato} campeonato - Dados do campeonato
     * @returns {Promise<Object>} Resposta da API
     */
    async criarCampeonato(campeonato) {
        return this.request('/campeonatos', {
            method: 'POST',
            body: JSON.stringify(campeonato.toJSON())
        });
    }

    /**
     * Lista todos os campeonatos
     * @returns {Promise<Campeonato[]>} Lista de campeonatos
     */
    async listarCampeonatos() {
        const data = await this.request('/campeonatos');
        return data.map(campeonatoData => Campeonato.fromJSON(campeonatoData));
    }

    /**
     * Atualiza um campeonato
     * @param {string} nomeOriginal - Nome original do campeonato
     * @param {Campeonato} campeonato - Novos dados do campeonato
     * @returns {Promise<Object>} Resposta da API
     */
    async atualizarCampeonato(nomeOriginal, campeonato) {
        return this.request(`/campeonatos?nome_original=${nomeOriginal}`, {
            method: 'PUT',
            body: JSON.stringify(campeonato.toJSON())
        });
    }

    /**
     * Deleta um campeonato
     * @param {string} nome - Nome do campeonato
     * @returns {Promise<Object>} Resposta da API
     */
    async deletarCampeonato(nome) {
        return this.request(`/campeonatos?nome=${nome}`, {
            method: 'DELETE'
        });
    }

    /**
     * Adiciona um clube a um campeonato
     * @param {string} nomeCampeonato - Nome do campeonato
     * @param {string} siglaClube - Sigla do clube
     * @returns {Promise<Object>} Resposta da API
     */
    async adicionarClubeAoCampeonato(nomeCampeonato, siglaClube) {
        return this.request(`/campeonatos/adicionar_clube?nome_campeonato=${nomeCampeonato}&sigla_clube=${siglaClube}`, {
            method: 'POST'
        });
    }

    /**
     * Lista clubes participantes de um campeonato
     * @param {string} nome - Nome do campeonato
     * @returns {Promise<Clube[]>} Lista de clubes participantes
     */
    async listarClubesCampeonato(nome) {
        const data = await this.request(`/campeonatos/clubes_participantes?nome=${nome}`);
        return data.map(clubeData => Clube.fromJSON(clubeData));
    }

    // ==================== PARTIDAS ====================

    /**
     * Registra uma nova partida
     * @param {Partida} partida - Dados da partida
     * @returns {Promise<Object>} Resposta da API
     */
    async registrarPartida(partida) {
        return this.request('/partidas', {
            method: 'POST',
            body: JSON.stringify(partida.toJSON())
        });
    }
}

File: trabalho_3/cliente_java/src/services/ApiService.java
package services;

import models.Clube;
import models.Campeonato;
import models.Partida;
import utils.JsonUtils;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * Serviço para comunicação com a API do servidor
 */
public class ApiService {
    private final String baseUrl;

    public ApiService() {
        this("http://localhost:8000");
    }

    public ApiService(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    // ==================== MÉTODOS HTTP BÁSICOS ====================

    private ApiResponse sendPost(String endpoint, String jsonBody) {
        try {
            URL url = new URL(baseUrl + endpoint);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json; utf-8");
            conn.setDoOutput(true);

            if (jsonBody != null && !jsonBody.isEmpty()) {
                try (OutputStream os = conn.getOutputStream()) {
                    byte[] input = jsonBody.getBytes("utf-8");
                    os.write(input, 0, input.length);
                }
            }

            return processResponse(conn);
        } catch (Exception e) {
            return new ApiResponse(false, "Erro: " + e.getMessage(), "");
        }
    }

    private ApiResponse sendGet(String endpoint) {
        try {
            URL url = new URL(baseUrl + endpoint);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            return processResponse(conn);
        } catch (Exception e) {
            return new ApiResponse(false, "Erro: " + e.getMessage(), "");
        }
    }

    private ApiResponse sendPut(String endpoint, String jsonBody) {
        try {
            URL url = new URL(baseUrl + endpoint);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("PUT");
            conn.setRequestProperty("Content-Type", "application/json; utf-8");
            conn.setDoOutput(true);

            try (OutputStream os = conn.getOutputStream()) {
                byte[] input = jsonBody.getBytes("utf-8");
                os.write(input, 0, input.length);
            }

            return processResponse(conn);
        } catch (Exception e) {
            return new ApiResponse(false, "Erro: " + e.getMessage(), "");
        }
    }

    private ApiResponse sendDelete(String endpoint) {
        try {
            URL url = new URL(baseUrl + endpoint);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("DELETE");
            return processResponse(conn);
        } catch (Exception e) {
            return new ApiResponse(false, "Erro: " + e.getMessage(), "");
        }
    }

    private ApiResponse processResponse(HttpURLConnection conn) throws IOException {
        int responseCode = conn.getResponseCode();
        boolean success = responseCode >= 200 && responseCode < 300;

        InputStream inputStream = success ? conn.getInputStream() : conn.getErrorStream();
        String response = "";
        
        if (inputStream != null) {
            Scanner scanner = new Scanner(inputStream).useDelimiter("\\A");
            response = scanner.hasNext() ? scanner.next() : "";
            scanner.close();
        }

        conn.disconnect();
        return new ApiResponse(success, success ? "Sucesso" : "Erro (" + responseCode + ")", response);
    }

    // ==================== MÉTODOS DE CLUBES ====================

    public ApiResponse criarClube(Clube clube) {
        return sendPost("/clubes", clube.toJson());
    }

    public ApiResponse<List<Clube>> listarClubes() {
        ApiResponse response = sendGet("/clubes");
        if (response.isSuccess()) {
            List<Clube> clubes = new ArrayList<>();
            String data = (String) response.getData();
            List<String> clubeJsons = JsonUtils.extractJsonObjects(data);
            for (String clubeJson : clubeJsons) {
                clubes.add(Clube.fromJson(clubeJson));
            }
            return new ApiResponse<>(true, "Sucesso", clubes);
        }
        return new ApiResponse<>(false, response.getMessage(), new ArrayList<>());
    }

    public ApiResponse<Clube> obterEstatisticasClube(String sigla) {
        ApiResponse response = sendGet("/clubes/estatisticas?sigla=" + sigla);
        if (response.isSuccess()) {
            String data = (String) response.getData();
            Clube clube = Clube.fromJson(data);
            return new ApiResponse<>(true, "Sucesso", clube);
        }
        return new ApiResponse<>(false, response.getMessage(), null);
    }

    public ApiResponse atualizarClube(String siglaOriginal, Clube clube) {
        try {
            String endpoint = "/clubes?sigla_original=" + URLEncoder.encode(siglaOriginal, "UTF-8");
            return sendPut(endpoint, clube.toJson());
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse(false, "Erro de codificação: " + e.getMessage(), "");
        }
    }

    public ApiResponse deletarClube(String sigla) {
        try {
            String endpoint = "/clubes?sigla=" + URLEncoder.encode(sigla, "UTF-8");
            return sendDelete(endpoint);
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse(false, "Erro de codificação: " + e.getMessage(), "");
        }
    }

    public ApiResponse<List<Partida>> obterHistoricoClube(String sigla) {
        try {
            String endpoint = "/clubes/partidas?sigla=" + URLEncoder.encode(sigla, "UTF-8");
            ApiResponse response = sendGet(endpoint);
            if (response.isSuccess()) {
                List<Partida> partidas = new ArrayList<>();
                String data = (String) response.getData();
                List<String> partidaJsons = JsonUtils.extractJsonObjects(data);
                for (String partidaJson : partidaJsons) {
                    partidas.add(Partida.fromJson(partidaJson));
                }
                return new ApiResponse<>(true, "Sucesso", partidas);
            }
            return new ApiResponse<>(false, response.getMessage(), new ArrayList<>());
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse<>(false, "Erro de codificação: " + e.getMessage(), new ArrayList<>());
        }
    }

    // ==================== MÉTODOS DE CAMPEONATOS ====================

    public ApiResponse criarCampeonato(Campeonato campeonato) {
        return sendPost("/campeonatos", campeonato.toJson());
    }

    public ApiResponse<List<Campeonato>> listarCampeonatos() {
        ApiResponse response = sendGet("/campeonatos");
        if (response.isSuccess()) {
            List<Campeonato> campeonatos = new ArrayList<>();
            String data = (String) response.getData();
            List<String> campeonatoJsons = JsonUtils.extractJsonObjects(data);
            for (String campeonatoJson : campeonatoJsons) {
                campeonatos.add(Campeonato.fromJson(campeonatoJson));
            }
            return new ApiResponse<>(true, "Sucesso", campeonatos);
        }
        return new ApiResponse<>(false, response.getMessage(), new ArrayList<>());
    }

    public ApiResponse atualizarCampeonato(String nomeOriginal, Campeonato campeonato) {
        try {
            String endpoint = "/campeonatos?nome_original=" + URLEncoder.encode(nomeOriginal, "UTF-8");
            return sendPut(endpoint, campeonato.toJson());
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse(false, "Erro de codificação: " + e.getMessage(), "");
        }
    }

    public ApiResponse deletarCampeonato(String nome) {
        try {
            String endpoint = "/campeonatos?nome=" + URLEncoder.encode(nome, "UTF-8");
            return sendDelete(endpoint);
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse(false, "Erro de codificação: " + e.getMessage(), "");
        }
    }

    public ApiResponse adicionarClubeAoCampeonato(String nomeCampeonato, String siglaClube) {
        try {
            String endpoint = "/campeonatos/adicionar_clube?nome_campeonato=" + 
                            URLEncoder.encode(nomeCampeonato, "UTF-8") + 
                            "&sigla_clube=" + URLEncoder.encode(siglaClube, "UTF-8");
            return sendPost(endpoint, "");
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse(false, "Erro de codificação: " + e.getMessage(), "");
        }
    }

    public ApiResponse<List<Clube>> listarClubesCampeonato(String nome) {
        try {
            String endpoint = "/campeonatos/clubes_participantes?nome=" + URLEncoder.encode(nome, "UTF-8");
            ApiResponse response = sendGet(endpoint);
            if (response.isSuccess()) {
                List<Clube> clubes = new ArrayList<>();
                String data = (String) response.getData();
                List<String> clubeJsons = JsonUtils.extractJsonObjects(data);
                for (String clubeJson : clubeJsons) {
                    clubes.add(Clube.fromJson(clubeJson));
                }
                return new ApiResponse<>(true, "Sucesso", clubes);
            }
            return new ApiResponse<>(false, response.getMessage(), new ArrayList<>());
        } catch (UnsupportedEncodingException e) {
            return new ApiResponse<>(false, "Erro de codificação: " + e.getMessage(), new ArrayList<>());
        }
    }

    // ==================== MÉTODOS DE PARTIDAS ====================

    public ApiResponse registrarPartida(Partida partida) {
        return sendPost("/partidas", partida.toJson());
    }

    // ==================== CLASSE INTERNA PARA RESPOSTA ====================

    public static class ApiResponse<T> {
        private final boolean success;
        private final String message;
        private final T data;

        public ApiResponse(boolean success, String message, T data) {
            this.success = success;
            this.message = message;
            this.data = data;
        }

        public boolean isSuccess() { return success; }
        public String getMessage() { return message; }
        public T getData() { return data; }

        @Override
        public String toString() {
            return String.format("ApiResponse{success=%s, message='%s', data=%s}", 
                               success, message, data);
        }
    }
}
